core/drivers/crypto/caam/cipher/caam_cipher_mac.c

```diff
@@ -0,0 +1,850 @@
+							&dst_align,
+							ctx->alg->size_ctx,
+							&realloc);
+		if (retstatus != CAAM_NO_ERROR) {
+			CIPHER_TRACE("Dest buffer reallocation error");
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out_free;
+		}
+	}
+
+	/* Check if there is some data saved to complete the buffer */
+	if (ctx->blockbuf.filled) {
+		srcbuf.length = ctx->alg->size_block - ctx->blockbuf.filled;
+
+		if (dupdate->last)
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, &dst_align,
+						      CIPHER_BLOCK_IN);
+		else
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, NULL,
+						      CIPHER_BLOCK_IN);
+
+		ctx->blockbuf.filled = 0;
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_GENERIC;
+			goto out_free;
+		}
+
+		size_todo -= ctx->alg->size_block;
+
+		if (size_todo || size_topost)
+			inc_mac_buffer(&srcbuf, srcbuf.length);
+	}
+
+	srcbuf.length = ctx->alg->size_block;
+
+	while (size_todo) {
+		if (dupdate->last)
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, &dst_align,
+						      CIPHER_BLOCK_NONE);
+		else
+			retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+						      true, &srcbuf, NULL,
+						      CIPHER_BLOCK_NONE);
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_GENERIC;
+			goto out_free;
+		}
+
+		size_todo -= ctx->alg->size_block;
+
+		if (size_todo || size_topost)
+			inc_mac_buffer(&srcbuf, srcbuf.length);
+	};
+
+	if (dst_align.data) {
+		if (!dst_align.nocache)
+			cache_operation(TEE_CACHEINVALIDATE, dst_align.data,
+					dst_align.length);
+
+		if (realloc)
+			memcpy(dupdate->dst.data, dst_align.data,
+			       MIN(dupdate->dst.length, dst_align.length));
+	}
+
+out:
+	ret = TEE_SUCCESS;
+
+	if (size_topost) {
+		struct caambuf cpysrc = {
+			.data = srcbuf.data,
+			.length = size_topost
+		};
+
+		CIPHER_TRACE("Save input data %zu bytes of %zu", size_topost,
+			     dupdate->src.length);
+
+		retstatus = caam_cpy_block_src(&ctx->blockbuf, &cpysrc, 0);
+		if (retstatus != CAAM_NO_ERROR)
+			ret = TEE_ERROR_GENERIC;
+	}
+
+out_free:
+	if (realloc)
+		caam_free_buf(&dst_align);
+
+	return ret;
+}
+
+/*
+ * Build and run the CMAC descriptor (AES only)
+ *
+ * @ctx     Cipher data context
+ * @srcbuf  Input data
+ * @dstbuf  [out] Output data if last block
+ * @last    Last block flag
+ */
+static TEE_Result run_cmac_desc(struct cipherdata *ctx, struct caambuf *srcbuf,
+				struct caambuf *dstbuf, bool last)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	struct caam_jobctx jobctx = { };
+	uint32_t *desc = NULL;
+	struct caamsgtbuf src_sgt = { .sgt_type = false };
+	struct caamsgtbuf dst_sgt = { .sgt_type = false };
+
+	desc = ctx->descriptor;
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+
+	if (last) {
+		retstatus = caam_sgt_build_block_data(&dst_sgt, NULL, dstbuf);
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_GENERIC;
```

Comment:
minor: I think here `return TEE_ERROR_MEMORY;` would be more accurate.

Reply:
I prefer to keep ```TEE_ERROR_GENERIC``` since ```caam_sgt_build_block_data()``` can return something else than ```CAAM_OUT_OF_MEMORY```

Reply:
sorry, I meant using `return` rather than `ret = ...; goto out;`.
`return TEE_ERROR_GENERIC`.
Sorry my typo, i cpoy/pasted the wrong error code.

Reply:
Ok, no problem.
I changed it to ```return TEE_ERROR_GENERIC```